### PR TITLE
[RSI, performance] stop the heartbeat/scheduler fan-out to every worker, client, and artifact path

### PR DESCRIPTION
**Pull request by snimu**
No-Ticket: RSI performance follow-up from the static perf review of the daemon fleet; no tracked ticket.

## Problem

Supervisor heartbeat/scheduler hot paths fan out linearly (or worse) with fleet size. Evidence: static perf review of the daemon fleet and persistence I/O (persistence-io.md F7/F8, daemon-fleet.md F1/F9), all verified on current main before the work:

- **heartbeats_list fan-out**: every global heartbeats_list request forwarded to EVERY live ready worker (`packages/coding-agent/src/modes/daemon/daemon-supervisor.ts:2490-2493` on base); every agents view polls every 15s (`modes/agents-view/agents-view-mode.ts:107`) and re-lists on every heartbeats_changed - a C x W round-trip amplifier.
- **Fleet-wide scheduled-jobs rescans**: every daemon-owned heartbeat/cron mutation and every worker residency change dropped the shared snapshot and re-scanned the ENTIRE session fleet (ledger family read + per-session store reads; `daemon-supervisor.ts:989-994`, `1096-1145`, `7103-7112` on base), and every open heartbeats stream re-scanned at least every 5s (PASSIVE_SCHEDULED_JOBS_REFRESH_MS).
- **Unfiltered broadcasts**: every mutation wrote heartbeats_changed to ALL connected clients (`daemon-supervisor.ts:7109-7111` on base), including one-shot prompt and viewer clients.
- **Cron-store mutation blast radius**: every mutation took a proper-lockfile lock on EVERY registered artifact path, ELOCKED retries busy-waited with Atomics.wait (up to 1s of event-loop blocking per path), and each changed file got 2 fsyncs (content + directory) (`packages/coding-agent/src/core/cron-jobs.ts:1565-1608`, `1676-1688`, `862-948` on base).

## Change

Four parts, one commit each:

1. **Serve heartbeats_list from per-worker snapshots** - each live worker's last complete heartbeat snapshot is the serving source; a heartbeats_changed frame refreshes only the reporting worker (only a worker heartbeats_changed frame queues the trailing refresh pass; lists just join the in-flight refresh), missing or frame-marked snapshots forward to that worker alone, and snapshots older than 60s refresh in the background (stale-while-revalidate) so a missed frame cannot pin stale data forever. Lists still fail rather than serving a partial catalog.
2. **Drive scheduled wakes from the aggregate snapshot** - the wake recompute and due-wake passes read the shared snapshot instead of rescanning the fleet; daemon-owned passive manage/cancel patch the mutated job's row in place, a worker gaining residency drops its now-covered rows in memory (re-derived from the current snapshot after the ledger read; a failed drop invalidates and forces a fresh scan), and worker frames do no passive work at all. Worker stops and saved-session delete/rename keep the invalidating full rescan, pending ephemeral-cancel intents keep the fresh-enumeration cadence, and a stale-while-revalidate refresh that finds changed rows re-arms the wake timer (external-writer convergence). cron_list serves the snapshot too.
3. **heartbeats_changed only to tracking clients, coalesced** - a client becomes a tracking client on its first scheduled-job command or by attaching with the new `heartbeat_catalog` client capability (the ACP adapter advertises it: `pi acp` attaches without ever issuing a scheduled-job command but still acts on the frame by publishing session_info_update to IDEs). Bursts collapse into one trailing-edge write per 100ms window; the timer clears on shutdown. Wire-behavior change: passive third-party daemon listeners must declare `heartbeat_catalog` (or issue a scheduled-job command) to keep receiving the push.
4. **Narrow cron-store locks, single fsync, non-blocking retries** - every store mutation is recomputed from the state read while the artifact lock is held: a lock-free probe finds the paths the mutation changes, the mutation is recomputed under those paths' locks, and a probe that found nothing locks all registered paths (same policy as mutateStates) so a concurrent writer in the probe-to-lock gap is not missed. The updatedAt-based freshness merge inside writeJobs is gone (it survives only in the legacy-to-artifact migration). Lock acquisition uses proper-lockfile's async lock() with the same bounded 100x10ms retry now awaited instead of Atomics.wait busy-waited, and writes keep a single content fsync (fsyncDir dropped: losing only the rename rolls the file back to the previous valid state, which interrupted-dispatch recovery already tolerates). All store mutators are async; the scheduler arms an immediate first claim and runs startup recovery as the first step of that claim; the daemon's sessionReplaced rebind is awaited before the session_replaced broadcast; the ephemeral-cancel tree walk and the passive heartbeat_manage re-check worker coverage inside the store's locked section (details under Review fixes below).

Also: daemon-side passive-job lookups now resolve paused and cancelled rows (base's active-only fresh scans skipped them), so managing/cancelling a paused passive heartbeat resolves instead of missing.

### Review fixes (commit d1559a21b)

"fix(coding-agent): recompute cron-store mutations under the lock and close the scheduler races" - 5 files, +536/-473 (src/core/cron-jobs.ts, src/modes/daemon/daemon-supervisor.ts, and the three test files below):

- Every catalog mutation - create, heartbeat create, pause/resume/clear/manage, cancel, per-session cancel, rebind, RLM heartbeat create/update/delete/cancel, record run/skip result - is recomputed from the state read while the artifact lock is held instead of reusing a pre-lock snapshot. Two overlapping heartbeat_set commands for one session now leave exactly one active heartbeat instead of both staying active, and the old updatedAt-based freshness merge inside writeJobs is gone (it survives only in the legacy-to-artifact migration).
- A no-op store mutation now takes locks as well: when the probe finds nothing to change, all registered artifact paths are locked and the mutation is recomputed under them, matching the policy mutateStates already had; the legacy single-file store no longer rewrites the file on a no-op.
- The scheduler's start() arms an immediate first claim, and startup recovery of interrupted dispatches runs as the first step of that claim, marked done only on success. A failing recovery rejects that claim exactly like a failing claimDue and is retried by the next claim; before, the failure was swallowed and the stale dispatch record made claimDue skip the job until restart (a scheduler that was never started still performs no recovery, and stop() during the recovery await aborts the claim).
- heartbeats_list callers that join an in-flight worker snapshot refresh no longer queue a trailing refresh pass; only a worker heartbeats_changed frame does, so a list no longer adds one extra worker RPC per joiner.
- The ephemeral tree cancel uses one predicate - the intent is still wanted and no tree member is covered by a resident worker, the same findWorkerBySessionFile check the scan uses, wrapped in a new sessionFileCovered helper - for the pre-check, the per-iteration exit (now tree-wide instead of per-session), and as a stillWanted guard the store re-evaluates inside its locked critical section. A promotion or worker registration that lands while a cancel waits for the artifact lock keeps the schedules.
- heartbeat_manage on a passive job re-checks under the artifact lock that neither the job's root session file nor its own session file is covered by a resident worker; if one is, the artifact stays untouched and the command is forwarded to that worker, whose own scheduler wakes. Downside: a worker that is still launching can answer the forward with an error such as "Unknown active session" or "No active heartbeat found" instead of the previous silent passive write, and intermediate ancestors are not checked because that would need a ledger read on this path.
- pauseHeartbeat, resumeHeartbeat and clearHeartbeat select the latest live heartbeat under the lock, using shared latestHeartbeat/isLiveHeartbeatFor helpers with getHeartbeat.

Follow-up (intentionally out of scope): passivated-session heartbeat fires still fully rehydrate the session file (`session-manager.ts:1493-1530`, `daemon-mode.ts:2472` on base); a folded-context snapshot for heartbeat-triggered rehydration is a separate change with its own risk profile.

Documented boundary: with zero attached clients, an externally written scheduled-job artifact converges at the next invalidating event (worker stop, saved-session delete/rename) or the next snapshot-refresh-then-re-arm; the re-arm path needs catalog reads.

Adapted to main's quota-resume store usage after #2375.

## Tests/Validation

All runs in a Prime sandbox (node:24), repo synced at the commit:

- `npx tsgo --noEmit -p tsconfig.json` (repo root): exit 0.
- `npx biome check --error-on-warnings packages/coding-agent`: exit 0.
- Vitest, 13 suites, 389/389 passed: cron-jobs (56), daemon-supervisor-eviction (35), daemon-supervisor-heartbeats (19), daemon-mode (72), agent-connection-daemon (92), agent-session-retry-events (60), interactive-heartbeat-management (17), cron-jobs-state-snapshot (11), proper-lockfile-compromise (8), 4257-update-restart-resume (10), heartbeats-list-response-latency (1), 4527-worker-heartbeat-scheduling (1), 2259-scheduled-catalog-metadata (7).
- check-test-policy: PASS locally against base 18a56bf35.
- Tests from the earlier commits cover the changed behavior end to end: repeated lists do not fan out; only the reporting worker refreshes; bursts coalesce (frames -> one refresh pass, mutations -> one broadcast); aged snapshots serve while refreshing; the wake timer arms from the snapshot with NO scan and residency gains drop covered rows without a scan; a failed drop forces a fresh scan; daemon-owned manage patches the snapshot without a rescan; a promotion landing mid-cancel (between cancels, or while a cancel waits for the artifact lock) keeps the sessions' jobs; a deferred sessionReplaced blocks the broadcast and a throwing one rejects the awaited rebind; the attach request carries heartbeat_catalog only when tracking; a held lock on an untouched path no longer stalls a mutation; the contended-lock retry yields the event loop; every scheduled-jobs write is fsync-once with no fsyncDir.

Six added tests, each shown failing on the previous head before the fix:

- cron-jobs: "keeps one heartbeat when two creates for a session overlap" - two overlapping creates left ['active','active'] before the fix.
- cron-jobs: "keeps a claimed dispatch with its job when another catalog write lands" - artifact-mode dispatch retention through a catalog write.
- cron-jobs: "retries a failed startup recovery before the next claim instead of skipping the stale dispatch" - runDue resolved 0 instead of rejecting before the fix.
- daemon-supervisor-eviction: "keeps schedules when a worker covers the tree while the cancel waits for the artifact lock" - ended with ['cancelled'] before the fix.
- daemon-supervisor-heartbeats: "lets concurrent lists join one refresh without queuing a trailing pass" - forwardToWorker was called 2 times before the fix.
- daemon-supervisor-heartbeats: "forwards a manage to the worker that became resident while the passive write waited for the lock" - no forward and ['paused'] before the fix.

Three deleted tests, all named "when a stale snapshot is written", plus the writeJobsForTest private-API helper: they exercised the freshness merge through the private writeJobs list API, which no longer exists.

Known pre-existing flake (verified failing identically on base e311d6495 by stash-and-rerun): daemon-supervisor-process.test.ts "exits an orphaned session worker when no replacement supervisor can come up" (worker supervisor-lost watchdog path, untouched by this PR).

## Coverage deliberately not restored

- The three deleted stale-snapshot tests: the private-API scenario they exercised is no longer reachable.
- heartbeat_manage during the launch window of an intermediate ancestor session: not covered by the new guard and not covered by a test.
- cron_cancel's passive branch keeps its lookup-then-write shape: a late cancel is read by the worker's next claim and has no unarmed-timer consequence; unchanged here, no test added.
- The daemon-level "process exits on an unhandled rejection from the timer-driven claim" channel: pre-existing behavior for claimDue failures, no test added.

---

> [!NOTE]
> **Medium Risk**
> Touches daemon scheduling, cross-process cron persistence, and client-visible push semantics (clients must opt into `heartbeats_changed`); missed awaits on new async store APIs would cause silent persistence races.
> 
> **Overview**
> Reduces supervisor and cron-store amplification when the fleet or client count grows.
> 
> **Daemon supervisor:** Global `heartbeats_list` no longer forwards to every worker on each request—it serves **per-worker cached snapshots**, refreshes only the reporting worker on `heartbeats_changed`, and background-refreshes snapshots older than 60s. Scheduled wake recomputes and `cron_list` lean on a **shared passive scheduled-jobs snapshot** (in-place row patches on daemon mutations, residency drops covered rows without a full scan; invalidating events still force rescans). **`heartbeats_changed` is coalesced (100ms) and sent only to clients that track the catalog** (scheduled-job commands or attach with `heartbeat_catalog`, e.g. ACP). Passive manage/cancel can resolve **paused/cancelled** jobs via snapshot-first lookup.
> 
> **Cron job store:** Mutators are **async** with **async file locks** and timer-based retries instead of blocking `Atomics.wait`. Writes **lock only artifact paths that actually change** (probe pass, then locked apply) and use **one content fsync** per write. Scheduler startup waits for interrupted-dispatch recovery before claiming due jobs.
> 
> **Session/runtime:** Quota-park durable wakes and RLM heartbeat host CRUD **await** cron-store I/O; daemon session rebind/replace and teardown paths **await** cron operations before broadcasting or continuing close logic.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5e9f1487af93c406970cb05386da1748ea1de246. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Stop heartbeat/scheduler fan-out with cached worker snapshots and targeted notifications
> - `DaemonSupervisor` now caches per-worker heartbeat snapshots (60s max age) with single-flight refresh, so `heartbeats_list` serves from cache instead of forwarding to every worker on each call
> - `heartbeats_changed` pushes are sent only to clients that issued a tracking command or declared the `heartbeat_catalog` capability, coalesced into a 100ms trailing window
> - Passive scheduled-job lookups use the cached snapshot first, falling back to a disk scan only on miss; worker residency removes covered passive rows without a full rescan
> - `AgentCronJobStore` mutations and file locks converted to async so contended locks retry with timers instead of blocking the event loop; `writeJobs` and `mutateStates` lock only artifact paths whose state changes
> - Behavioral Change: `heartbeats_list` now reports a worker as a failure when it has neither a usable cached snapshot nor a ready connection, instead of returning a partial result; cron state writes no longer request a directory fsync; `sessionReplaced` callback in `ActiveSessionBindingCallbacks` may now be async and is awaited before broadcasting `session_replaced`
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 5e9f148.</sup>
> 
>

Comments minimized per review.

### CODE REVIEW

File: packages/coding-agent/src/modes/daemon/daemon-mode.ts [hunk 1]
```diff
@@ -2206,19 +2210,19 @@ export class AgentDaemon {
...
-	): AgentCronJob {
+	): Promise<AgentCronJob> {
 		const session = state.runtime.session;
 		const sessionFile = session.sessionFile;
 		if (!sessionFile) {
 			throw new Error("Heartbeats require a persisted session file");
 		}
 		const previousHeartbeat = this.cronStore.getHeartbeat(state.activeSessionId);
-		const job = this.cronStore.createHeartbeat({
+		const job = await this.cronStore.createHeartbeat({
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/daemon/daemon-mode.ts:2225]: Valid, acknowledged; leaving this open. `createHeartbeat` reads the catalog at cron-jobs.ts:337 and only then `await this.writeJobs(...)` at :370; `writeJobs` yields at :934 inside `withCronJobsStateLocks`, and `mergeJobsStates` (:1858-1889) keeps current jobs that are absent from the incoming list, so two overlapping in-process creates for one session can persist two active heartbeats. The command path is fire-and-forget (`void this.handleLine`, daemon-mode.ts:3676, 3698), so the window is real. In `main` the store was synchronous, so this is introduced by the async conversion. The file lock cannot serialize in-process callers; the fix belongs in `AgentCronJobStore` (one in-process mutation chain around the read-modify-write methods), which is more than a triage-sized change, so it is reported for a follow-up rather than patched here.

*[written by prime-agent; thread-resolution run approved by snimu]*

File: packages/coding-agent/src/core/cron-jobs.ts [hunk 2]
```diff
@@ -356,7 +367,7 @@ export class AgentCronJobStore {
 			nextRunAt: parsed.nextRunAt.toISOString(),
 			runCount: 0,
 		};
-		this.writeJobs([...existing, job]);
+		await this.writeJobs([...existing, job]);
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/cron-jobs.ts]: Valid, acknowledged; leaving this open. Same finding as the Macroscope thread at daemon-mode.ts:2225: read at cron-jobs.ts:337, lock taken only inside `await this.writeJobs` (:370, :934), and `mergeJobsStates` (:1866-1869) retains rows absent from a stale incoming list. Needs an in-process serialization at the store layer (all `readJobs()` -> `await writeJobs()` methods share the shape); reported for a follow-up.

*[written by prime-agent; thread-resolution run approved by snimu]*

File: packages/coding-agent/src/modes/daemon/daemon-mode.ts [hunk 3]
```diff
@@ -1783,8 +1783,8 @@ export class AgentDaemon {
 			this.bindingCompletions.delete(state.activeSessionId);
 			completeBinding();
 		}
-		this.registerCronStoreForState(state);
-		this.rebindCronJobsToState(state);
+		await this.registerCronStoreForState(state);
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/daemon/daemon-mode.ts:1786]: Pre-existing, not introduced here; resolving. In `main`, `addRuntime` has the identical structure: the cleanup `try/catch` ends at daemon-mode.ts:1785 and `registerCronStoreForState` / `rebindCronJobsToState` follow it at :1786-1787 as synchronous calls, so a throw there (e.g. `lockSync` ELOCKED, artifact I/O) escaped the cleanup boundary in the same way. This PR only made those two calls `await`ed; it did not move them. Moving them inside the boundary is also not a drop-in change (`bindingSessions` still holds the id inside the boundary, which alters `runCronJob`'s half-bound fallback at :2454-2459), so it is out of scope for a perf PR.

*[written by prime-agent; thread-resolution run approved by snimu]*

File: packages/coding-agent/src/modes/daemon/daemon-supervisor.ts [hunk 4]
```diff
@@ -1145,9 +1262,140 @@ export class DaemonSupervisor {
...
+		return (await this.collectPassiveScheduledJobs()).find(({ job }) => match(job));
+	}
+
+	/**
+	 * Daemon-owned passive scheduled-job mutation (manage/cancel on a
+	 * non-resident session): patch the mutated job's row in the aggregate
+	 * snapshot instead of dropping the whole snapshot and rescanning the
+	 * fleet, then re-arm the wake timer from the patched rows.
+	 */
+	private onPassiveScheduledJobMutated(job: AgentCronJob): void {
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/daemon/daemon-supervisor.ts:1323]: Not reachable, resolving. `scanPassiveScheduledJobs` (daemon-supervisor.ts:1216-1263) has its last `await` at :1230; every artifact read after that is synchronous (`store.list()` -> `readFileSync`, :1245-1261) and `startSharedPassiveScheduledJobsScan` publishes in the `.then` at :1084-1086, so "read every artifact" and "publish" are one event-loop turn. The mutation side spans real I/O: `store.manageHeartbeat`/`store.cancel` (:2796, :2866) write and `await release()` the lock (cron-jobs.ts:1659-1662) before `onPassiveScheduledJobMutated` runs (:2798, :2868). A scan's read+publish turn therefore lands either entirely before the write (its rows are then patched by :1389) or after it (it reads the mutated artifact). The "read before write, publish after patch" interleaving cannot occur.

*[written by prime-agent; thread-resolution run approved by snimu]*

File: packages/coding-agent/src/modes/daemon/daemon-supervisor.ts [hunk 5]
```diff
@@ -1145,9 +1262,140 @@ export class DaemonSupervisor {
...
+		const cached = this.passiveScheduledJobs;
+		if (cached) {
+			this.passiveScheduledJobs = {
+				...cached,
+				rows: cached.rows.map((row) => (row.job.id === job.id ? { ...row, job } : row)),
+			};
+		}
+		this.scheduleScheduledSessionWakeRecompute();
+		this.notifyHeartbeatsChanged();
+	}
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/daemon/daemon-supervisor.ts:1333]: Not reachable, resolving. Same analysis as the Macroscope thread at :1384: the scan's artifact reads and its publish are one synchronous turn after the last `await` at daemon-supervisor.ts:1230 (:1245-1261, :1084-1086), while the mutation's write and lock release complete before the patch is applied (:2796-2798, :2866-2868; cron-jobs.ts:1659-1662). For `dropPassiveRowsCoveredByWorkers` (:1315-1327), the scan computes coverage live in the same synchronous block (:1256, :1276), so it can only publish covered rows if it ran before residency was gained, in which case the later drop patches those rows; the author documents that ordering at :1309-1313.

*[written by prime-agent; thread-resolution run approved by snimu]*

File: packages/coding-agent/src/core/cron-jobs.ts [hunk 6]
```diff
@@ -1562,47 +1608,58 @@ function isDueJob(job: AgentCronJob, now: Date): boolean {
...
+ * Serialize a read-modify-write against other processes for exactly the given
+ * paths. Acquisition and the ELOCKED retry are async (proper-lockfile lock()),
+ * so contention never blocks the event loop; the critical section itself runs
+ * synchronously once every lock is held.
+ */
+async function withCronJobsStateLocks<T>(paths: readonly string[], action: () => T): Promise<T> {
+	const releases: Array<() => Promise<void>> = [];
+	try {
+		for (const path of [...new Set(paths)].sort()) {
+			releases.push(await acquireCronJobsLock(path));
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/cron-jobs.ts:1616]: Not reachable in the daemon, resolving. Startup `recoverInterruptedDispatches` runs only from `AgentCronScheduler.start` (cron-jobs.ts:1047-1048). In worker mode the store is `forSessionArtifacts()` (daemon-mode.ts:644-645) and `start()` runs at :738, before any `addRuntime` registers an artifact (:1821-1833), so `paths` is empty at cron-jobs.ts:858 and the call is a no-op; per-artifact recovery happens in `recoverSessionArtifact` (:221-235), which locks that one path and applies without a probe. In legacy mode there is exactly one path, so "lock A, miss B" cannot arise. There is also no second claimer: `claimDue` is called only by the owning worker's scheduler (:1091); the supervisor's stores never claim. The only multi-path recovery call is in `test/cron-jobs.test.ts:1351`.

*[written by prime-agent; thread-resolution run approved by snimu]*

File: packages/coding-agent/src/modes/daemon/daemon-supervisor.ts [hunk 7]
```diff
@@ -1145,9 +1262,140 @@ export class DaemonSupervisor {
...
+	 */
+	private async findPassiveScheduledJob(
+		match: (job: AgentCronJob) => boolean,
+	): Promise<PassiveScheduledJob | undefined> {
+		const cached = this.passiveScheduledJobs;
+		if (cached) {
+			const row = cached.rows.find(({ job }) => match(job));
+			if (row) return row;
+		}
+		return (await this.collectPassiveScheduledJobs()).find(({ job }) => match(job));
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/daemon/daemon-supervisor.ts:1319]: Misread of the filter, resolving. `collectPassiveScheduledJobs()` defaults to `activeScheduledJobs(rows)` (daemon-supervisor.ts:1024-1028), and that filter keeps `status === "active" || status === "paused"` (:467-469), so paused passive heartbeats are found by the fresh-scan fallback. Only `cancelled`/`completed` rows are dropped, and for those both mutations are no-ops (`manageHeartbeat` returns the job unchanged, cron-jobs.ts:611-613; `cancel` skips cancelled, :643-645), so finding them would reach the same undefined branch. `main` used the very same `collectPassiveScheduledJobs()` call (no `true`) at daemon-supervisor.ts:2549 and :2621.

*[written by prime-agent; thread-resolution run approved by snimu]*

File: packages/coding-agent/src/core/cron-jobs.ts [hunk 8]
```diff
@@ -1676,17 +1733,63 @@ function writeJobsFile(path: string, jobs: readonly AgentCronJob[], mergeCurrent
 function writeJobsState(path: string, state: CronJobsState): CronJobsFileIdentity | undefined {
 	mkdirSync(dirname(path), { recursive: true, mode: 0o700 });
 	let written: CronJobsFileIdentity | undefined;
+	// One fsync per write: the temp file's contents are durable before the
+	// rename, which itself is atomic. Losing the rename only after a power
+	// failure rolls the mutation back to the previous valid file, which the
+	// interrupted-dispatch recovery already tolerates.
 	writeFileAtomicSync(path, `${JSON.stringify(state, null, 2)}\n`, {
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/cron-jobs.ts:1697]: Overstated, resolving. `writeFileAtomicSync` renames the temp file onto the existing path (atomic-file.ts:72). If a power loss drops the rename, the old directory entry survives and the previous valid state is read; the file is not missing, so `readState` does not fall back to `EMPTY_JOBS_STATE` (except for a first-ever write, whose previous state was already "no file"). The residual effect is a rollback of the last mutation on power loss only (a process crash keeps the rename in the page cache), which the comment at cron-jobs.ts:1736-1739 states and which dispatch recovery tolerates. Whether a freshly created job must survive power loss is a product decision; if so the change is `fsyncDir: true` at :1742, noted in the triage report for the maintainer.

*[written by prime-agent; thread-resolution run approved by snimu]*

File: packages/coding-agent/src/core/cron-jobs.ts [hunk 9]
```diff
@@ -720,65 +735,34 @@ export class AgentCronJobStore {
...
+			dispatches: [] as AgentCronDispatch[],
+			results: recordDispatchResultInState(state, dispatchId, result),
+		}));
+		return results[0];
 	}
 
-	recoverInterruptedDispatches(now = new Date()): AgentCronJob[] {
-		const recovered: AgentCronJob[] = [];
-		this.mutateStates((state) => {
+	async recoverInterruptedDispatches(now = new Date()): Promise<AgentCronJob[]> {
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/cron-jobs.ts:750]: Valid, acknowledged; leaving this open. `.catch(() => undefined)` at cron-jobs.ts:1049-1050 swallows a failed startup recovery, `hasStarted` stays true, and `claimDue` skips jobs with an existing dispatch record (:1795, :1805), so a stale dispatch stays unclaimable until restart; the comment at :1049 ("surfaces again on the next claim") is not accurate. In `main` the call was synchronous and a failure propagated out of daemon start. Scope is narrow: in worker mode no artifact is registered when `start()` runs, so recovery is a no-op there; this affects the legacy single-file store on lock timeout or disk error. The fix needs a small design decision (retry recovery from `runDue` with a floor vs. propagate), so it is reported for a follow-up rather than patched here.

*[written by prime-agent; thread-resolution run approved by snimu]*

File: packages/coding-agent/src/modes/daemon/daemon-mode.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jimmyarphs2) [reply packages/coding-agent/src/modes/daemon/daemon-mode.ts:2225]: Confirmed and fixed. createHeartbeat previously read before the asynchronous lock, allowing overlapping replacements to build from the same stale catalog. Heartbeat replacement now runs through the locked state mutator and recomputes from the current state after lock acquisition. Added a concurrent #2487 regression; the focused suite, shuffled runs, and repository checks pass.

[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/daemon/daemon-mode.ts:2225]: Fixed in d1559a21b. Every `AgentCronJobStore` mutator now hands `writeJobs` a catalog mutator that is recomputed from the state read while the artifact lock is held, instead of a list computed before the lock; the `updatedAt` freshness merge that let a stale list survive that recompute is removed. The lock already serializes conforming writers, in-process included (`proper-lockfile` returns `ELOCKED` to a second same-process `lock()`), so an overlapping second `createHeartbeat` now sees and cancels the first one. Regression test: "keeps one heartbeat when two creates for a session overlap" in test/cron-jobs.test.ts. (The earlier "Confirmed and fixed" reply above did not come from this branch.)

*[written by prime-agent; thread-resolution run approved by snimu]*

File: packages/coding-agent/src/core/cron-jobs.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/cron-jobs.ts:750]: Fixed in d1559a21b. Startup recovery is no longer fired-and-swallowed from `start()`. It runs as the first step of `runDue` behind a `recovered` latch that is set only on success, so a failed recovery rejects that claim exactly like a failed `claimDue` does (from the timer that reaches the daemon's unhandled-rejection handler, as on main) and is retried by the next claim; no claim runs over an unrecovered store. `start()` now just arms an immediate first claim. Regression test: "retries a failed startup recovery before the next claim instead of skipping the stale dispatch" in test/cron-jobs.test.ts.

*[written by prime-agent; thread-resolution run approved by snimu]*

File: packages/coding-agent/src/modes/daemon/daemon-supervisor.ts [hunk 10]
```diff
@@ -7039,7 +7215,15 @@ export class DaemonSupervisor {
...
-			store.cancelJobsForSession({ sessionFile });
+			// Each cancel yields on a lock, so a promotion can commit between cancels;
+			// re-check intent and coverage so a stale intent never kills a live worker's schedules.
+			if (stillWanted && !stillWanted()) return;
+			try {
+				if (this.findWorkerBySessionFile(sessionFile, exclude)) return;
+			} catch {
+				return;
+			}
+			await store.cancelJobsForSession({ sessionFile });
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/daemon/daemon-supervisor.ts]: Fixed in d1559a21b. `cancelJobsForSession` now takes the cancel intent as a `stillWanted` guard that the store re-evaluates inside its locked, synchronous critical section. `cancelScheduledJobsForSessionTree` passes one predicate (intent still wanted and no tree member covered, via the same `findWorkerBySessionFile` check the scan uses) as pre-check, per-iteration exit and lock-time guard, so a promotion or worker registration that commits while the cancel waits for the artifact lock leaves the schedules untouched. Regression test: "keeps schedules when a worker covers the tree while the cancel waits for the artifact lock" in test/daemon-supervisor-eviction.test.ts.

*[written by prime-agent; thread-resolution run approved by snimu]*

File: packages/coding-agent/src/modes/daemon/daemon-supervisor.ts [hunk 11]
```diff
@@ -1079,6 +1109,87 @@ export class DaemonSupervisor {
...
+	}
+
+	/**
+	 * Targeted single-flight refresh of one worker: frames and lists join the
+	 * in-flight pass; a mid-refresh frame queues one trailing pass.
+	 */
+	private refreshWorkerHeartbeatSnapshot(worker: ResidentWorker): Promise<void> {
+		const inFlight = worker.heartbeatSnapshotRefresh;
+		if (inFlight) {
+			worker.heartbeatSnapshotRefreshQueued = true;
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/daemon/daemon-supervisor.ts]: Fixed in d1559a21b: only `onWorkerHeartbeatsChanged` (the worker-frame path) queues the trailing pass now; joining lists just await the in-flight refresh, so continuous polling no longer issues paired RPCs. One correction on the impact: joined callers were never pending until transport timeout. The flag is cleared at the top of each loop iteration and each pass is one 5s-bounded forward, so a join cost at most one extra bounded pass. Regression test: "lets concurrent lists join one refresh without queuing a trailing pass" in test/daemon-supervisor-heartbeats.test.ts.

*[written by prime-agent; thread-resolution run approved by snimu]*

File: packages/coding-agent/src/modes/daemon/daemon-supervisor.ts [hunk 12]
```diff
@@ -2542,18 +2718,18 @@ export class DaemonSupervisor {
 				);
 				if (!cachedWorker) {
 					// Passive jobs are managed against their durable store; no wake just to flip a status.
-					const passive = (await this.collectPassiveScheduledJobs()).find(
-						({ job }) => job.id === command.jobId && job.activeSessionId === command.activeSessionId,
+					const passive = await this.findPassiveScheduledJob(
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/daemon/daemon-supervisor.ts:2721]: Fixed in d1559a21b. `manageHeartbeat` now accepts a `stillWanted` guard evaluated with the artifact locked; the passive branch passes "the row's root session file and the job's session file are still uncovered" (same `findWorkerBySessionFile` predicate as the scan). If a worker became resident during the lookup or the lock wait, the store leaves the artifact unchanged and the existing code path forwards the command to that worker, whose own scheduler wakes. A re-check placed only after the lookup would still have left the lock-wait window open, which is why the check lives in the store's critical section. Regression test: "forwards a manage to the worker that became resident while the passive write waited for the lock" in test/daemon-supervisor-heartbeats.test.ts.

*[written by prime-agent; thread-resolution run approved by snimu]*

**Review by sethkarten — APPROVED** [review]
Design review passed: claim-vs-diff verified, revert-probe coverage checked, edge cases and open threads assessed. LGTM.

— Prime Agent (sethkarten's agent)
